src/learner.cc

```diff
@@ -741,13 +748,19 @@ class LearnerImpl : public Learner {
       metrics_.back()->Configure({cfg_.begin(), cfg_.end()});
     }
     for (size_t i = 0; i < data_sets.size(); ++i) {
-      DMatrix * dmat = data_sets[i];
-      this->ValidateDMatrix(dmat);
-      this->PredictRaw(data_sets[i], &preds_[dmat], false);
-      obj_->EvalTransform(&preds_[dmat]);
+      std::shared_ptr<DMatrix> dmat = data_sets[i];
+      auto& predt = this->cache_->Cache(dmat, generic_parameters_.gpu_id);
+      this->ValidateDMatrix(dmat.get());
+      this->PredictRaw(data_sets[i].get(), &predt, false);
+
+      predictions_.SetDevice(generic_parameters_.gpu_id);
+      predictions_.Resize(predt.predictions.Size());
+      predictions_.Copy(predt.predictions);
```

Comment:
Is there an extra copy here that wasn't there before?

Reply:
@RAMitchell It's a copied moved from `Predictor::PredictFromCache`.  I removed this function altogether as now the cache is managed by `Learner`, hence the copying is also done by `Learner`.  It's 1 copy less as I also removed the temporary vector in C API.